File: 백준/Gold/14461. 소가 길을 건너간 이유 7/소가 길을 건너간 이유 7.cc
#include <bits/stdc++.h>
#include <ext/rope>
#define fastio cin.tie(0), cout.tie(0), ios::sync_with_stdio(0);
#define all(x) (x).begin(), (x).end()
#define x first 
#define y second
using namespace std; using ll = long long;
using ld = long double; using pld = pair<ld, ld>;
using i128 = __int128_t; using f128 = __float128; 
using pll = pair<ll, ll>; using tll = tuple<ll, ll, ll>;
ll n, m, k, t = 1; string s;

constexpr ll INF = 0x3f3f3f3f3f3f3f3f;
constexpr ll MINF = 0xc0c0c0c0c0c0c0c0;
constexpr ll MAX = 101; // SET MAX SIZE
constexpr ll MOD = 998244353;
ll a[MAX][MAX];
ll dy[4] = {0, 0, 1, -1}, dx[4] = {1, -1, 0, 0};

bool outrange(ll cy, ll cx){
    return cy < 0 || cx < 0 || cy >= n || cx >= n;
}

class _dij {
public:
    class node{
    public:
        ll d;
        node() : node(0){}
        node(ll d) : d(d){}
        ll num() const{ return d; }
        bool operator<(const node& ot) const{ return num() < ot.num(); }
        bool operator>(const node& ot) const{ return num() > ot.num(); }
        bool operator==(const node& ot) const{ return num() == ot.num(); }
        bool operator<=(const node& ot) const{ return num() <= ot.num(); }
        node operator+(const node& ot) const{
            return {d + ot.d};
        }
        operator ll(){ return d; }
    };
    node mx(){ return {INF}; }
    node mn(){ return {0}; }

    using ptl = pair <node, ll>;
    ll n; vector <node> d;
    vector <ll> pre;
    vector <vector<ptl>> adj;
    priority_queue <ptl, vector<ptl>, greater<ptl>> pq;

    _dij(){}
    _dij(ll n) { this->n = n; adj.resize(n + 1); }

    void add(ll st, ll en, node c) { // 양방향
        adj[st].push_back({ c,en });
        adj[en].push_back({ c,st });
    }
    void addsol(ll st, ll en, node c) { // 단방향
        adj[st].push_back({ c,en });
    }

    void init(ll st) {
        d.clear(); pre.clear();
        d.resize(n + 1, mx()); pre.resize(n + 1, -1); 
        pq.push({ mn(), st });
        d[st] = mn();

        while (!pq.empty()) {
            auto [cn, cur] = pq.top(); pq.pop();
            if(cn > d[cur]) continue;
            
            for (auto& i : adj[cur]) {
                auto [nn, nxt] = i;
                node pl = nn + cn;
        
                if (d[nxt] <= pl) continue;
                d[nxt] = pl; pre[nxt] = cur; 
                pq.push({ pl, nxt });
            }
        }
    }

    node ret(ll n) { return d[n]; }
};

ll num(ll cy, ll cx, ll cnt){
    return cnt * n * n + cy * n + cx;
}

void run(){
    cin >> n >> m; _dij dij(3 * (n + 1) * (n + 1));
    for(int i = 0;i < n;i++){
        for(int j = 0;j < n;j++) cin >> a[i][j];
    }

    for(int i = 0;i < n;i++){
        for(int j = 0;j < n;j++){
            for(int k = 0;k < 4;k++){
                ll ny = i + dy[k], nx = j + dx[k];
                if(outrange(ny, nx)) continue;
                dij.addsol(num(i, j, 0), num(ny, nx, 1), m);
                dij.addsol(num(i, j, 1), num(ny, nx, 2), m);
                dij.addsol(num(i, j, 2), num(ny, nx, 0), m + a[ny][nx]);
            }
        }
    }

    dij.init(num(0, 0, 0));
    cout << min({dij.ret(num(n - 1, n - 1, 0)), dij.ret(num(n - 1, n - 1, 1)), dij.ret(num(n - 1, n - 1, 2))});
}

int main() {
    fastio; // cin >> t;
    while(t--) run(); 

    return 0;
}

File: 백준/Gold/14461. 소가 길을 건너간 이유 7/소가 길을 건너간 이유 7.py
import sys
import heapq

INF = 10**12
MAX = 3 * 10201

input = sys.stdin.readline

n, m = map(int, input().split())
a = [[0] * (n + 2) for _ in range(n + 2)]
for i in range(1, n + 1):
    row = list(map(int, input().split()))
    for j in range(1, n + 1):
        a[i][j] = row[j - 1]

adj = [[] for _ in range(MAX)]
d = [INF] * MAX
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def outrange(cy, cx):
    return cy <= 0 or cx <= 0 or cy > n or cx > n

def num(cy, cx, cnt):
    return cnt * 10101 + cy * 101 + cx

for i in range(1, n + 1):
    for j in range(1, n + 1):
        for k in range(4):
            ny = i + dy[k]
            nx = j + dx[k]
            if outrange(ny, nx):
                continue
            adj[num(i, j, 0)].append((num(ny, nx, 1), m))
            adj[num(i, j, 1)].append((num(ny, nx, 2), m))
            adj[num(i, j, 2)].append((num(ny, nx, 0), m + a[ny][nx]))

pq = []
heapq.heappush(pq, (0, num(1, 1, 0)))

while pq:
    cd, cur = heapq.heappop(pq)
    if d[cur] <= cd:
        continue
    d[cur] = cd
    for nxt, co in adj[cur]:
        if d[nxt] <= cd + co:
            continue
        heapq.heappush(pq, (cd + co, nxt))

result = INF
for i in range(3):
    result = min(result, d[num(n, n, i)])
print(result)
